pkg/mtbroker/filter/filter_handler.go

```diff
@@ -80,93 +84,132 @@ func (e sendError) Unwrap() error {
 		return
 	}
-	writer.WriteHeader(http.StatusOK)
-}
 
-// Start begins to receive messages for the handler.
-//
-// Only HTTP POST requests to the root path (/) are accepted. If other paths or
-// methods are needed, use the HandleRequest method directly with another HTTP
-// server.
-//
-// This method will block until a message is received on the stop channel.
-func (r *Handler) Start(ctx context.Context) error {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
-	errCh := make(chan error, 1)
-	go func() {
-		errCh <- r.ceClient.StartReceiver(ctx, r.serveHTTP)
-	}()
-
-	// We are ready.
-	r.isReady.Store(true)
-
-	// Stop either if the receiver stops (sending to errCh) or if stopCh is closed.
-	select {
-	case err := <-errCh:
-		return err
-	case <-ctx.Done():
-		break
+	// Check if the event should be sent.
+	filterResult := h.shouldSendEvent(ctx, &t.Spec, event)
+
+	if filterResult == failFilter {
+		h.Logger.Debug("Event did not pass filter", zap.Any("triggerRef", triggerRef))
+		// We do not count the event. The event will be counted in the broker ingress.
+		// If the filter didn't pass, it means that the event wasn't meant for this Trigger.
+		writer.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// Record the event processing time. This might be off if the receiver and the filter pods are running in
+	// different nodes with different clocks.
+	var arrivalTimeStr string
+	if extErr := event.ExtensionAs(broker.EventArrivalTime, &arrivalTimeStr); extErr == nil {
+		arrivalTime, err := time.Parse(time.RFC3339, arrivalTimeStr)
+		if err == nil {
+			_ = h.Reporter.ReportEventProcessingTime(reportArgs, time.Since(arrivalTime))
+		}
 	}
 
-	// No longer ready.
-	r.isReady.Store(false)
+	target := subscriberURI.String()
+	req, err := h.Sender.NewCloudEventRequestWithTarget(ctx, target)
+	if err != nil {
+		h.Logger.Info("failed to create the request", zap.String("target", target))
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	req.Header = utils.PassThroughHeaders(request.Header)
+	err = cehttp.WriteRequest(ctx, binding.ToMessage(event), req)
+	if err != nil {
+		h.Logger.Info("failed to write request")
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
 
-	// stopCh has been closed, we need to gracefully shutdown h.ceClient. cancel() will start its
-	// shutdown, if it hasn't finished in a reasonable amount of time, just return an error.
-	cancel()
-	select {
-	case err := <-errCh:
-		return err
-	case <-time.After(writeTimeout):
-		return errors.New("timeout shutting down ceClient")
+	start := time.Now()
+	resp, err := h.Sender.Send(req)
+	dispatchTime := time.Since(start)
+	if err != nil {
+		h.Logger.Error("failed to dispatch message", zap.Error(err))
+		_ = h.Reporter.ReportEventDispatchTime(reportArgs, http.StatusInternalServerError, dispatchTime)
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	// TODO response might contain another event
```

Comment:
so you want to go from response to event right?

```
responseMessage := http.NewMessageFromHttpResponse(resp)
responseEvent, err := binding.ToEvent(responseMessage)
```

Reply:
and then where should this event be sent?

Reply:
ah, i have no clue :smile: where the event was sent previously?

Reply:
ah ok i see it was sent back as response, so what you should do to fill the response writer with the event is:

```
http.WriteResponseWriter(context, responseWriter, statusCode, responseMessage)
```

Reply:
I don't know :confused: 
The SDK uses reflection in that part and it's not easy to understand what it's doing.

Reply:
Cool! thank you! :+1: